File: sdk/python/src/zrok2/listing.py
from zrok_api.api import MetadataApi
from zrok2.environment.root import Root
import zrok2.model as model


def list_shares(root: Root, **filters) -> list[model.ShareDetail]:
    """List shares with optional filters.

    Analogous to ``zrok2 list shares``.

    Args:
        root: The Root environment context.
        **filters: Optional keyword filters passed to the API:
            env_zid, share_mode, backend_mode, share_token, target,
            permission_mode, has_activity, idle, activity_duration,
            created_after, created_before, updated_after, updated_before.

    Returns:
        List of ShareDetail objects.
    """
    if not root.IsEnabled():
        raise Exception("environment is not enabled; enable with 'zrok enable' first!")

    try:
        zrok = root.Client()
    except Exception as e:
        raise Exception("error getting zrok client", e)

    try:
        metadata_api = MetadataApi(zrok)
        res = metadata_api.list_shares(**filters)
    except Exception as e:
        raise Exception("unable to list shares", e)

    shares = res.shares or []
    return [
        model.ShareDetail(
            Token=s.share_token or "",
            ZId=s.z_id or "",
            EnvZId=s.env_zid or "",
            ShareMode=s.share_mode or "",
            BackendMode=s.backend_mode or "",
            FrontendEndpoints=s.frontend_endpoints or [],
            Target=s.target or "",
            Limited=s.limited or False,
            CreatedAt=s.created_at or 0,
            UpdatedAt=s.updated_at or 0,
        )
        for s in shares
    ]


def list_accesses(root: Root, **filters) -> list[model.AccessDetail]:
    """List accesses with optional filters.

    Analogous to ``zrok2 list accesses``.

    Args:
        root: The Root environment context.
        **filters: Optional keyword filters passed to the API:
            env_zid, share_token, bind_address, description,
            created_after, created_before, updated_after, updated_before.

    Returns:
        List of AccessDetail objects.
    """
    if not root.IsEnabled():
        raise Exception("environment is not enabled; enable with 'zrok enable' first!")

    try:
        zrok = root.Client()
    except Exception as e:
        raise Exception("error getting zrok client", e)

    try:
        metadata_api = MetadataApi(zrok)
        res = metadata_api.list_accesses(**filters)
    except Exception as e:
        raise Exception("unable to list accesses", e)

    accesses = res.accesses or []
    return [
        model.AccessDetail(
            Id=a.id or 0,
            FrontendToken=a.frontend_token or "",
            EnvZId=a.env_zid or "",
            ShareToken=a.share_token or "",
            BackendMode=a.backend_mode or "",
            BindAddress=a.bind_address or "",
            Description=a.description or "",
            Limited=a.limited or False,
            CreatedAt=a.created_at or 0,
            UpdatedAt=a.updated_at or 0,
        )
        for a in accesses
    ]

File: sdk/nodejs/sdk/src/zrok/listing.ts
import {Root} from "./environment";
import {MetadataApi} from "../api";
import {ShareDetail, AccessDetail} from "./model";

export interface ListSharesFilters {
    envZId?: string;
    shareMode?: string;
    backendMode?: string;
    shareToken?: string;
    target?: string;
    permissionMode?: string;
    hasActivity?: boolean;
    idle?: boolean;
    activityDuration?: string;
    createdAfter?: string;
    createdBefore?: string;
    updatedAfter?: string;
    updatedBefore?: string;
}

export interface ListAccessesFilters {
    envZId?: string;
    shareToken?: string;
    bindAddress?: string;
    description?: string;
    createdAfter?: string;
    createdBefore?: string;
    updatedAfter?: string;
    updatedBefore?: string;
}

export const listShares = async (root: Root, filters?: ListSharesFilters): Promise<ShareDetail[]> => {
    if (!root.isEnabled()) {
        throw new Error("environment is not enabled; enable with 'zrok2 enable' first!");
    }

    const cfg = await root.client();
    const res = await new MetadataApi(cfg).listShares(filters || {})
        .catch(err => {
            throw new Error("unable to list shares: " + err);
        });

    const shares = res.shares || [];
    return shares.map(s => ({
        token: s.shareToken || "",
        zId: s.zId || "",
        envZId: s.envZId || "",
        shareMode: s.shareMode || "",
        backendMode: s.backendMode || "",
        frontendEndpoints: s.frontendEndpoints || [],
        target: s.target || "",
        limited: s.limited || false,
        createdAt: s.createdAt || 0,
        updatedAt: s.updatedAt || 0,
    }));
}

export const listAccesses = async (root: Root, filters?: ListAccessesFilters): Promise<AccessDetail[]> => {
    if (!root.isEnabled()) {
        throw new Error("environment is not enabled; enable with 'zrok2 enable' first!");
    }

    const cfg = await root.client();
    const res = await new MetadataApi(cfg).listAccesses(filters || {})
        .catch(err => {
            throw new Error("unable to list accesses: " + err);
        });

    const accesses = res.accesses || [];
    return accesses.map(a => ({
        id: a.id || 0,
        frontendToken: a.frontendToken || "",
        envZId: a.envZId || "",
        shareToken: a.shareToken || "",
        backendMode: a.backendMode || "",
        bindAddress: a.bindAddress || "",
        description: a.description || "",
        limited: a.limited || false,
        createdAt: a.createdAt || 0,
        updatedAt: a.updatedAt || 0,
    }));
}
